controller/gke-cluster-config-handler.go

```diff
@@ -0,0 +1,349 @@
+package controller
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	gkev1 "github.com/rancher/gke-operator/pkg/apis/gke.cattle.io/v1"
+	v12 "github.com/rancher/gke-operator/pkg/generated/controllers/gke.cattle.io/v1"
+	"github.com/rancher/rke/log"
+	wranglerv1 "github.com/rancher/wrangler/pkg/generated/controllers/core/v1"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/google"
+	gkeapi "google.golang.org/api/container/v1"
+	"google.golang.org/api/option"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+const (
+	controllerName           = "gke-controller"
+	controllerRemoveName     = "gke-controller-remove"
+	gkeConfigCreatingPhase   = "creating"
+	gkeConfigNotCreatedPhase = ""
+	gkeConfigActivePhase     = "active"
+	gkeConfigUpdatingPhase   = "updating"
+	gkeConfigImportingPhase  = "importing"
+	allOpen                  = "0.0.0.0/0"
+	gkeClusterConfigKind     = "GKEClusterConfig"
+	runningStatus            = "RUNNING"
+	none                     = "none"
+)
+
+type Handler struct {
+	gkeCC           v12.GKEClusterConfigClient
+	gkeEnqueueAfter func(namespace, name string, duration time.Duration)
+	gkeEnqueue      func(namespace, name string)
+	secrets         wranglerv1.SecretClient
+	secretsCache    wranglerv1.SecretCache
+}
+
+func Register(
+	ctx context.Context,
+	secrets wranglerv1.SecretController,
+	gke v12.GKEClusterConfigController) {
+
+	controller := &Handler{
+		gkeCC:           gke,
+		gkeEnqueue:      gke.Enqueue,
+		gkeEnqueueAfter: gke.EnqueueAfter,
+		secretsCache:    secrets.Cache(),
+		secrets:         secrets,
+	}
+
+	// Register handlers
+	gke.OnChange(ctx, controllerName, controller.recordError(controller.OnGkeConfigChanged))
+	gke.OnRemove(ctx, controllerRemoveName, controller.OnGkeConfigRemoved)
+}
+
+func (h *Handler) OnGkeConfigChanged(key string, config *gkev1.GKEClusterConfig) (*gkev1.GKEClusterConfig, error) {
+	if config == nil {
+		return nil, nil
+	}
+
+	if config.DeletionTimestamp != nil {
+		return nil, nil
+	}
+
+	switch config.Status.Phase {
+	case gkeConfigImportingPhase:
+		fmt.Errorf("not implemented")
+	case gkeConfigNotCreatedPhase:
+		return h.create(config)
+	case gkeConfigCreatingPhase:
+		return h.waitForCreationComplete(config)
+	case gkeConfigActivePhase:
+		return h.checkAndUpdate(config)
+	case gkeConfigUpdatingPhase:
+		fmt.Errorf("not implemented")
+	}
```

Comment:
Is there a chance that `config.Status.Phase` will be `nil` or that we need a `default` case here?

Reply:
I believe for that to be the case it would be a bug. There should be no `nil` or `default`. I thought about a default, but the default would be a noop and thus would be the same as not having it... though I could log at that point? so maybe good? Honestly we could have it fail in dev as it should never happen.